Add compatibility wrapper for DeepMind Melting Pot

elliottower opened the pull request:
See: https://github.com/deepmind/meltingpot/pull/117

Pytest and pre-commit hooks passing. 

Also fixed misc.pyrght errors (e.g., step() argument action of type int incompatible with int64 -> action: ActionType (from pettingzoo.utils.env) and made comments a bit more consistent (using the same language to describe the wrappers ,etc).

elliottower commented:
@jjshoots @pseudo-rnd-thoughts

pseudo-rnd-thoughts commented:
Could you fix pre-commit and tests

jjshoots commented on tests/test_meltingpot.py:
```diff
@@ -0,0 +1,57 @@
+
+@pytest.mark.parametrize("substrate_name", SUBSTRATES)
+def test_seeding(substrate_name):
+    """Tests the seeding of the openspiel conversion wrapper."""
```
Typo in this docstring.

elliottower replied:
Good catch

jjshoots commented on tests/test_meltingpot.py:
```diff
@@ -0,0 +1,57 @@
+"""Tests the functionality of the MeltingPotCompatibility wrapper on meltingpot substrates."""
```
The tests look good to me, but perhaps @pseudo-rnd-thoughts may want to have a second look.

jjshoots commented on shimmy/utils/meltingpot.py:
```diff
@@ -0,0 +1,61 @@
+"""Shared utils for meltingpot."""
```
This looks rather similar to the dm_env utils, do you think the two could be merged, especially the spec_to_space functionality? Or perhaps rename the timestep_to_observations into multiagent_timestep_to_observations to make it more verbose.

elliottower replied:
Good call I’ll look into that

jjshoots commented on shimmy/meltingpot_compatibility.py:
```diff
@@ -0,0 +1,193 @@
+        self.player_roles = substrate.get_config(self.substrate_name).default_player_roles
+        self.max_cycles = max_cycles
+        self.env_config = {"substrate": self.substrate_name, "roles": self.player_roles}
+        EzPickle.__init__(self, self.render_mode, self.env_config, self.max_cycles)
```
Could the ezpickle be moved up?

elliottower replied:
The EzPickle init line takes a few arguments which need to be extracted from the environment (env config, which contains player_roles) so it can't be done on the first line.

jjshoots reviewed (changes requested):
Overall LGTM, aside from some minor comments. @pseudo-rnd-thoughts may want to have a second look as he tends to be more detail oriented and may have preferences over certain aspects.

pseudo-rnd-thoughts commented on shimmy/utils/meltingpot.py:
```diff
@@ -0,0 +1,61 @@
+  })
+
+
+def spec_to_space(spec: tree.Structure[dm_env.specs.Array]) -> spaces.Space:
```
As jet mentioned, remove and replace with the existing dm spec to space

elliottower replied:
So the existing spec to space is actually less general than this one, and is a bit less clear (imo) so I’ve actually replaced the existing one with this and adapted the functionality to work with both cases. Will commit the changes later today.

pseudo-rnd-thoughts commented on tests/test_meltingpot.py:
```diff
@@ -0,0 +1,57 @@
+@pytest.mark.parametrize("substrate_name", SUBSTRATES)
+def test_passing_substrates(substrate_name):
+    """Tests the conversion of all openspiel envs."""
+    for _ in range(5):
```
Why are we running this 5 times?

elliottower replied:
I just mirrored the functionality of the openspiel tests, but maybe both should be removed as it does seem redundant and slow.

pseudo-rnd-thoughts replied:
@jjshoots Do you know why?

jjshoots replied:
On Openspiel this was needed because some of the envs weren't deterministic, and so randomizing it with 5 seeds can help reveal problems that only pop up sometimes.

pseudo-rnd-thoughts replied:
Ok, so we shouldn't have that problem in meltingpot

pseudo-rnd-thoughts commented on shimmy/meltingpot_compatibility.py:
```diff
@@ -0,0 +1,193 @@
+    ):
+        """Wrapper that converts a openspiel environment into a pettingzoo environment.
+
+        Args:
```
Is it too complex to add the capability to pass an environment if someone has made a custom environment? Otherwise, don't worry

elliottower replied:
I can look into that, they have a nice tutorial for making your own env so it would be nice to have.

elliottower replied:
To update this, I was referring to this tutorial they have on making custom envs: https://github.com/deepmind/meltingpot/blob/main/docs/substrate_tutorial/index.md
It's not clear how exactly you use an environment besides loading it explicitly through an import statement, and their link to an example game is dead so I can't look at that for reference. 

I think the best option for this kind of functionality would be to make something like `custom_environment_compatibility` which takes an already defined env (which you import directly at the top of the file, rather than passing substrate name to be built by a factory). I tried to implement it as you described with taking in `env` in the constructor, but either the `substrate_name` or `env` has to be a required arg, and I'm not sure how to make args mutually exclusive/only have one of the two required. I guess we could just have it be that you pass in a substrate_name anyways which isn't used, if you're loading from an existing env.

Here's my rough code idea for what you described:
```
    def __init__(
        self,
        substrate_name: str,
        render_mode: str | None,
        max_cycles: int = MAX_CYCLES,
        env: meltingpot.python.utils.substrates.substrate.Substrate = None,
    ):
        """Wrapper that converts a openspiel environment into a pettingzoo environment.

        Args:
            substrate_name (str): name of meltingpot substrate to load
            render_mode (Optional[str]): render_mode
            max_cycles (Optional[int]): maximum number of cycles (steps) before termination
            env (Optional[meltingpot.python.utils.substrates.substrate.Substrate]): existing meltingpot env to use

        """
        EzPickle.__init__(self, substrate_name, render_mode, max_cycles, env)

        if env is not None:
            self._env = env
        else:
            # Create env config
            self.substrate_name = substrate_name
            self.player_roles = meltingpot.python.substrate.get_config(
                self.substrate_name
            ).default_player_roles
            self.max_cycles = max_cycles
            self.env_config = {
                "substrate": self.substrate_name,
                "roles": self.player_roles,
            }

            # Build substrate from pickle
            self.env_config = config_dict.ConfigDict(self.env_config)
            self._env = meltingpot.python.substrate.build(
                self.env_config["substrate"], roles=self.env_config["roles"]
            )
```

pseudo-rnd-thoughts commented on tests/test_meltingpot.py:
```diff
@@ -0,0 +1,57 @@
+    a_space2.seed(42)
+
+    while env1.agents:
+        actions1 = {agent: env1.action_space(agent).sample() for agent in env1.agents}
```
Are all of the spaces shared otherwise this isn't deterministic.
@jjshoots Should each agent's space be a unique object?

jjshoots replied:
The way that PettingZoo is built, yes, each agent's space is a unique object. However, if we want to use a shared object for all of them, the way to do it is by making the function return the same action space regardless of the argument passed in. I believe this is already done [here](https://github.com/Farama-Foundation/Shimmy/pull/39/files/babbb09e7f93eb0a0005e70d322ffba9498c7829#diff-82ddd83028a0705b182c4f23382fb4b809c549d2ec6f694b21875b5de374828bR107).

elliottower replied:
Not sure what the best way to do this is, I can change it to use separate action spaces for each agent if that’s better. The previous melting pot code did the same thing as I’m doing now but I can see a rationale for making it separate per agent potentially. Would it be non deterministic because of how the functools caching works? I feel like since that’s the way I’ve seen it made consistent for over envs it probably makes sense to keep the action spaces the same. Just makes things simpler.

jjshoots replied:
Honestly if meltingpot has the same action spaces for each agent, this method is probably fine. I think an even faster way would be to do it [how KAZ does it](https://github.com/Farama-Foundation/PettingZoo/blob/master/pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py#L380). We eliminate the need for functools then. If all agents indeed share the same action space, we don't even need the dictionary, just return the same action space for every agent.

elliottower replied:
Defining them all in the constructor ahead of time sounds like a good idea to me although @pseudo-rnd-thoughts and you should make the decision I think. Is it faster to use dict(zip()) than using list comprehension in a dict? Style wise I feel like it's a little complex and not immediately obvious how it works with the unused variable _ and enumerate normally returning two values, etc. 
[KAZ:](https://github.com/Farama-Foundation/PettingZoo/blob/master/pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py#L380)
```
self.action_spaces = dict(
            zip(self.agents, [Discrete(6) for _ in enumerate(self.agents)])
        )
```
Dict comprehension:
```
self.action_spaces = {agent: Discrete(6) for agent in self.agents}
```

jjshoots replied:
Given that this is in the init, I doubt it will affect things too much as this is usually only called once per training session. That said, I agree that the second choice is much cleaner and we should probably do that.

pseudo-rnd-thoughts replied:
To me, I thought the ability for each individual agent to have their own unique action space allowing deterministic sampling for each agent would be a helpful property. That was the motivation for me with the proposal. 
@jjshoots Knows PZ better than I so I would follow his suggestion

pseudo-rnd-thoughts commented on tests/test_meltingpot.py:
```diff
@@ -0,0 +1,57 @@
+
+        returns1 = (observations1, rewards1, terminations1, truncations1)
+        returns2 = (observations2, rewards2, terminations2, truncations2)
+
```
Can we not just check that all of the data is equivalent? There is function in gymnasium `data_equivalence` that does this already

elliottower replied:
Good call I can replace it with that. Again this was mostly mirroring the code from the open spiel script but that means both can be updated to fix these minor issues.

jjshoots replied:
Good catch, if you could fix the openspiel tests using data_equivalence too, that'd be great.

jjshoots commented on shimmy/meltingpot_compatibility.py:
```diff
@@ -0,0 +1,193 @@
+        """
+        rgb_arr = self.state()[0]['WORLD.RGB']
+        if self.render_mode == 'human':
+            plt.cla()
```
True, I think we should avoid import matplotlib if possible. I'm not sure how it's done in meltingpot, but if it's not possible to not use matplotlib, I think that's still fine. We can simply make it a dependency for meltingpot only.

@elliottower is it possible to use PIL instead?

elliottower replied:
I’m not familiar with PIL but from some googling looks doable. https://stackoverflow.com/questions/57316491/how-to-convert-matplotlib-figure-to-pil-image-object-without-saving-image
Something like this? Or do you mean to avoid matplotlib entirely? Is it because plt is slow? And I’ve on seen pettingzoo games use pygame not PIL, is there a consensus on one or the other?

jjshoots replied:
We generally try to avoid imports of large libraries, especially if we're not using their main functionality. Since we only want to display images, MPL is a bit overkill. That said, PIL is also a bit of a large library, and so is PyGame. I think the consensus for one over the other has thus far just been 'most of the other libraries tend not to use MPL, so let's not use that'. There may be other reasons that I don't know of.

elliottower replied:
Makes sense, will look into it today.

elliottower commented on shimmy/meltingpot_compatibility.py:
```diff
@@ -0,0 +1,193 @@
+            self._env.observation_spec()[0]['WORLD.RGB'])  # type: ignore
+
+        # Set agents
+        self._num_players = len(self._env.observation_spec())
```
This was code from their previous pettingzoo code and the number of players stayed consistent with the observation and such, but I’ll check again to see if it might be an issue.

pseudo-rnd-thoughts replied:
What it an issue ?

elliottower commented on shimmy/meltingpot_compatibility.py:
```diff
@@ -0,0 +1,193 @@
+
+        state_space = utils.spec_to_space(
+            self._env.observation_spec()[0]['WORLD.RGB'])
+        return state_space
```
Do you mean to make it in the init constructor rather than a separate method? I had it that way originally but saw the action space and observation space in most other examples were separate methods so thought it might be more readable that way. Although those are functions with agent id as input, whereas this is independent of agent. Wasn’t sure if it was necessary at all because a lot of examples don’t have state space included.

pseudo-rnd-thoughts replied:
It is just confusing as the `observation-spec` doesn't change and you already define `self.state_space` which is equivalent to this function. So just use the variable that already exists or remove the variable

elliottower commented:
Addressed these comments (including the minor openspiel changes), cleaned up docstrings and swapped the rendering from plt to pygame. I also added a test for pygame rendering as it's different from how the underlying env does rendering by default, although it may not be necessary. 

Problem is now that the seeding seems to be broken and I'm not sure how to go about fixing it. @jjshoots @pseudo-rnd-thoughts

jjshoots commented:
@elliottower, seeding for meltingpot? It may be the case that meltingpot just has very bad seeding techniques that doesn't actually make it deterministic. This would not be the first time. Our solution to this was to set `nondeterministic=True` in the gymnasium API check, or disable seeding tests for those that are failing seeding in PZ tests.

elliottower commented:
> @elliottower, seeding for meltingpot? It may be the case that meltingpot just has very bad seeding techniques that doesn't actually make it deterministic. This would not be the first time. Our solution to this was to set `nondeterministic=True` in the gymnasium API check, or disable seeding tests for those that are failing seeding in PZ tests.

Oh okay that’s good to know. I asked how hard it would be to supply a seed and they said it wouldn’t be easy and recommended something similar to disable seeding.

pseudo-rnd-thoughts commented:
I found this in the CI 
```
----------------------------- Captured stderr call -----------------------------

Downloading CMU mocap data:   0%|          | 0.00/466M [00:00<?, ?it/s]
Downloading CMU mocap data:   1%|          | 4.00M/466M [00:00<00:21, 22.0Mit/s]
Downloading CMU mocap data:   3%|▎         | 16.0M/466M [00:00<00:08, 55.7Mit/s]
Downloading CMU mocap data:   5%|▌         | 24.0M/466M [00:00<00:08, 57.4Mit/s]
Downloading CMU mocap data:   9%|▊         | 40.0M/466M [00:00<00:05, 78.1Mit/s]
Downloading CMU mocap data:  12%|█▏        | 56.1M/466M [00:00<00:04, 95.9Mit/s]
Downloading CMU mocap data:  15%|█▌        | 72.1M/466M [00:00<00:03, 109Mit/s] 
Downloading CMU mocap data:  19%|█▉        | 88.1M/466M [00:01<00:03, 111Mit/s]
Downloading CMU mocap data:  23%|██▎       | 105M/466M [00:01<00:02, 128Mit/s] 
Downloading CMU mocap data:  26%|██▌       | 122M/466M [00:01<00:02, 141Mit/s]
Downloading CMU mocap data:  31%|███       | 143M/466M [00:01<00:02, 163Mit/s]
Downloading CMU mocap data:  34%|███▍      | 160M/466M [00:01<00:02, 156Mit/s]
Downloading CMU mocap data:  38%|███▊      | 175M/466M [00:01<00:02, 144Mit/s]
Downloading CMU mocap data:  41%|████▏     | [192](https://github.com/Farama-Foundation/Shimmy/actions/runs/4355883425/jobs/7613638215#step:4:193)M/466M [00:01<00:02, 129Mit/s]
Downloading CMU mocap data:  46%|████▋     | 216M/466M [00:01<00:01, 153Mit/s]
Downloading CMU mocap data:  50%|████▉     | 231M/466M [00:01<00:01, 155Mit/s]
Downloading CMU mocap data:  53%|█████▎    | 247M/466M [00:02<00:01, 124Mit/s]
Downloading CMU mocap data:  58%|█████▊    | 269M/466M [00:02<00:01, 150Mit/s]
Downloading CMU mocap data:  61%|██████▏   | 285M/466M [00:02<00:01, 127Mit/s]
Downloading CMU mocap data:  64%|██████▍   | 299M/466M [00:02<00:01, 116Mit/s]
Downloading CMU mocap data:  68%|██████▊   | 318M/466M [00:02<00:01, 133Mit/s]
Downloading CMU mocap data:  71%|███████▏  | 332M/466M [00:02<00:01, 107Mit/s]
Downloading CMU mocap data:  75%|███████▌  | 349M/466M [00:03<00:00, 122Mit/s]
Downloading CMU mocap data:  81%|████████  | 376M/466M [00:03<00:00, 158Mit/s]
Downloading CMU mocap data:  84%|████████▍ | 393M/466M [00:03<00:00, 134Mit/s]
Downloading CMU mocap data:  88%|████████▊ | 408M/466M [00:03<00:00, 96.7Mit/s]
Downloading CMU mocap data:  93%|█████████▎| 433M/466M [00:03<00:00, 126Mit/s] 
Downloading CMU mocap data:  96%|█████████▋| 449M/466M [00:03<00:00, 120Mit/s]
Downloading CMU mocap data: 100%|██████████| 466M/466M [00:03<00:00, 124Mit/s]
```

However, I don't seem to be able to replicate the issue

elliottower commented on shimmy/utils/dm_env.py:
```diff
@@ -44,17 +45,31 @@ def dm_spec2gym_space(spec) -> spaces.Space[Any]:
+    Array observations are converted to numpy arrays. Dict observations are converted recursively per key.
+
+    Args:
+      obs: The dm_env observation
```
Good catch

elliottower replied:
Is there no way to configure the pre-commit hooks to autoformat things like this?

elliottower commented on shimmy/utils/meltingpot.py:
Talked with @jjshoots about this a bit and it led to errors combining the two conversion functions so I separated them again. The other two helper functions in melting pot utils are specific to it as well (involving their observation’s world view keys).

jjshoots replied:
TBH, I think I prefer this technique of splitting the utils up into the various parts for dm env and meltingpot. It's very much single responsibility.

jjshoots commented on shimmy/openspiel_compatibility.py:
```diff
@@ -105,14 +110,20 @@ def action_space(self, agent: AgentID):
+
+        Print the current game state.
+        """
+        print(self.game_state)
```
This is kinda weird since some openspiel environments actually have a render function, but I guess c'est la vie.

elliottower replied:
I wasn’t actually aware of that, the current code just ignores rendering and says it’s not implemented but I could check if the env has a render function and if not just print the state.

jjshoots reviewed (approved):
Overall LGTM, some minor questions which I think once solved get my greenlight for merging.

elliottower commented on bin/install_dm_lab.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/sh
+set -eu
+
+if [[ "$(uname -s)" == 'Linux' ]]; then
```
Sounds like a good idea, do you agree @pseudo-rnd-thoughts ?

elliottower commented on tests/test_dm_control.py:
```diff
@@ -25,151 +25,151 @@
+#         "dm_control/compatibility-env-v0", env=wrapped_env, disable_env_checker=True
+#     )
+#     check_env(env.unwrapped, skip_render_check=True)
+#     env.close()
```
Oh this is definitely a mistake from my part, I thought for some reason when I synced with master that this changed but now looking at master that file hasn’t changed for a month. Will clean it up.

pseudo-rnd-thoughts reviewed (changes requested):
IMHO, this PR has become several different PRs: adding meltingpot support, documentation updates and possibly more given the number of files edited.
Therefore, I would propose closing this PR and opening new PRs for each of the problems addressed. This should make reviewing and understanding the PRs later if necessary

jjshoots commented:
@pseudo-rnd-thoughts Actually I think that's a bit counter productive. I mean for us it makes sense, but for Elliot it will be a whole lot of work

elliottower commented:
> @pseudo-rnd-thoughts Actually I think that's a bit counter productive. I mean for us it makes sense, but for Elliot it will be a whole lot of work

It’s all good, it’s best practice so I’m happy to practice it now and spend the extra day splitting things up and organizing. Looks messy to do a single PR commit with all these changes (although committing directly to master I feel like the sequential changes would be fine, it’s just that PRs get condensed into one commit afaik)

pseudo-rnd-thoughts commented:
@elliottower I would recommend not using main for prs as it can cause issues if you need to revert commits etc

elliottower commented on tests/test_openspiel.py:
```diff
@@ -148,8 +148,8 @@ def test_seeding(game):
 
-    agent1 = env1.agent_iter()
-    agent2 = env2.agent_iter()
+    agent1 = env1.agents[0]
+    agent2 = env2.agents[0]
```
I don't think this bit is a problem because it's just hard coding the first agent name in order to do the action space seeding, we still use the iter.
